processing/src/main/java/io/druid/query/ordering/StringComparators.java

```diff
@@ -342,16 +355,86 @@ public String toString()
     }
   }
 
-  public static StringComparator makeComparator(String type)
+  private static BigDecimal convertStringToBigDecimal(String input) {
+    // treat unparseable Strings as nulls
+    BigDecimal bd = null;
+    try {
+      bd = new BigDecimal(input);
+    } catch (NullPointerException | NumberFormatException ex) {
+    }
+    return bd;
+  }
+
+  public static class NumericComparator implements StringComparator
   {
-    if (type.equals(StringComparators.LEXICOGRAPHIC_NAME)) {
-      return LEXICOGRAPHIC;
-    } else if (type.equals(StringComparators.ALPHANUMERIC_NAME)) {
-      return ALPHANUMERIC;
-    } else if (type.equals(StringComparators.STRLEN_NAME)) {
-      return STRLEN;
-    } else {
-      throw new IAE("Unknown string comparator[%s]", type);
+    @Override
+    public int compare(String o1, String o2)
+    {
+      if (o1 == o2) {
+        return 0;
+      }
+
+      BigDecimal bd1 = convertStringToBigDecimal(o1);
+      BigDecimal bd2 = convertStringToBigDecimal(o2);
+
+      if (bd1 != null && bd2 != null) {
+        return bd1.compareTo(bd2);
+      }
+
+      if (bd1 == null && bd2 == null) {
+        // both Strings are unparseable, just compare lexicographically to have a well-defined ordering
+        return StringComparators.LEXICOGRAPHIC.compare(o1, o2);
+      }
+
+      if (bd1 == null) {
+        return -1;
+      } else {
+        return 1;
+      }
     }
+
+    @Override
+    public String toString()
+    {
+      return StringComparators.NUMERIC_NAME;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      return true;
+    }
+  }
+
+  public static StringComparator makeComparator(StringComparatorType type)
```

Comment:
I think we can actually get rid of `StringComparatorType` and hook this directly up to Jackson. Try annotating this with `@JsonCreator` and either putting it inside `StringComparator` (which would have to be an abstract class).

See QueryGranularity for an example of a class that works this way. It can accept both strings (like `"minute"` and objects (like `{"type":"period", ...}`)